make clip_grad_norm return python float

ngimel opened the pull request:
Differential Revision: D19388419

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ba81474d:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ba81474d4d5ad421bef7513b8544a7c6db70458c).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4233893) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (1/3)
**Step:** "Checkout pytorch/builder repo" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84858968) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84858968))
<details>
<summary>
<code>fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'.</code>
</summary>

```
+ [[ -n 32159 ]] 
+ git fetch --force origin pull/32159/head:remotes/origin/pull/32159 
remote: Enumerating objects: 8, done.         
remote: Counting objects:  75% (6/8)         remote: Counting objects:  87% (7/8)         remote: Counting objects: 100% (8/8)         remote: Counting objects: 100% (8/8), done.         
remote: Compressing objects:  50% (1/2)         remote: Compressing objects: 100% (2/2)         remote: Compressing objects: 100% (2/2), done.         
remote: Total 8 (delta 6), reused 7 (delta 6), pack-reused 0         
Unpacking objects:  75% (6/8)    Unpacking objects:  87% (7/8)    Unpacking objects: 100% (8/8)    Unpacking objects: 100% (8/8), done. 
From https://github.com/pytorch/pytorch 
 * [new ref]         refs/pull/32159/head -> origin/pull/32159 
+ git reset --hard ba81474d4d5ad421bef7513b8544a7c6db70458c 
fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4233915) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/3)
**Step:** "Checkout code" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84858975) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84858975))
<details>
<summary>
<code>fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'.</code>
</summary>

```
Receiving objects:  97% (308864/318416), 215.91 MiB | 33.24 MiB/s    Receiving objects:  98% (312048/318416), 215.91 MiB | 33.24 MiB/s    Receiving objects:  99% (315232/318416), 215.91 MiB | 33.24 MiB/s    remote: Total 318416 (delta 25), reused 19 (delta 3), pack-reused 318341         
Receiving objects:  99% (318395/318416), 215.91 MiB | 33.24 MiB/s    Receiving objects: 100% (318416/318416), 215.91 MiB | 33.24 MiB/s    Receiving objects: 100% (318416/318416), 229.30 MiB | 28.65 MiB/s, done. 
Resolving deltas:  98% (248158/253135)    Resolving deltas:  99% (250623/253135)    Resolving deltas: 100% (253135/253135)    Resolving deltas: 100% (253135/253135), done. 
remote: Enumerating objects: 8, done.         
remote: Counting objects:  75% (6/8)         remote: Counting objects:  87% (7/8)         remote: Counting objects: 100% (8/8)         remote: Counting objects: 100% (8/8), done.         
remote: Compressing objects:  50% (1/2)         remote: Compressing objects: 100% (2/2)         remote: Compressing objects: 100% (2/2), done.         
remote: Total 8 (delta 6), reused 7 (delta 6), pack-reused 0         
Unpacking objects:  75% (6/8)    Unpacking objects:  87% (7/8)    Unpacking objects: 100% (8/8)    Unpacking objects: 100% (8/8), done. 
From github.com:pytorch/pytorch 
 * [new ref]               refs/pull/32159/head -> origin/pull/32159 
fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4233912) pytorch_linux_xenial_py3_clang5_mobile_build (3/3)
**Step:** "Checkout code" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84858976) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84858976))
<details>
<summary>
<code>fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'.</code>
</summary>

```
Receiving objects: 100% (318416/318416), 224.66 MiB | 32.67 MiB/s    Receiving objects: 100% (318416/318416), 228.79 MiB | 30.02 MiB/s, done. 
Resolving deltas:  98% (248069/253074)    Resolving deltas:  99% (250556/253074)    Resolving deltas: 100% (253074/253074)    Resolving deltas: 100% (253074/253074), done. 
Warning: Permanently added the RSA host key for IP address '140.82.114.4' to the list of known hosts.  
remote: Enumerating objects: 8, done.         
remote: Counting objects:  75% (6/8)         remote: Counting objects:  87% (7/8)         remote: Counting objects: 100% (8/8)         remote: Counting objects: 100% (8/8), done.         
remote: Compressing objects:  50% (1/2)         remote: Compressing objects: 100% (2/2)         remote: Compressing objects: 100% (2/2), done.         
remote: Total 8 (delta 6), reused 7 (delta 6), pack-reused 0         
Unpacking objects:  75% (6/8)    Unpacking objects:  87% (7/8)    Unpacking objects: 100% (8/8)    Unpacking objects: 100% (8/8), done. 
From github.com:pytorch/pytorch 
 * [new ref]               refs/pull/32159/head -> origin/pull/32159 
fatal: Could not parse object 'ba81474d4d5ad421bef7513b8544a7c6db70458c'. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD commented on torch/nn/utils/clip_grad.py:
```diff
@@ -25,9 +25,12 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     norm_type = float(norm_type)
     if norm_type == inf:
-        total_norm = max(p.grad.detach().abs().max() for p in parameters)
+        total_norm = max(p.grad.detach().abs().max().item() for p in parameters) if len(parameters) != 0 else 0
```
Do you want this to be `0.` in the else to make it a float?

ngimel replied:
Yes, I'll make it a float. XLA is making it None in it's patch, I'm not sure if 0. or None makes more sense, but I'll stick with 0 for now.

albanD commented on torch/nn/utils/clip_grad.py:
```diff
@@ -25,9 +25,12 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     else:
-        total_norm = torch.norm(torch.stack([torch.norm(p.grad.detach(), norm_type) for p in parameters]), norm_type)
+        if len(parameters) == 0:
+            total_norm = 0
```
Same here, `0.` ?

albanD commented on torch/nn/utils/clip_grad.py:
```diff
@@ -25,9 +25,12 @@ def clip_grad_norm_(parameters, max_norm, norm_type=2):
     max_norm = float(max_norm)
     norm_type = float(norm_type)
     if norm_type == inf:
```
nit: Do we really want to keep that special case? We already have it in the implementation of norm.

ngimel replied:
Good point, I'll remove it.

albanD reviewed (approved):
LGTM. Small detail about `0` vs `0.` in python.

SsnL commented:
wasn't eliminating calling `.item()` one of the big points of the original PR?

ngimel commented:
> wasn't eliminating calling `.item()` one of the big points of the original PR?

a) Nope, the bad performance was attributed to .item() calls but they were not at fault
b) there was an .item() call per parameter, here it's a single .item() call per function

Commit:
make clip_grad_norm return python float

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/32159

Differential Revision: D19388419

fbshipit-source-id: a42a09b9fecace0a11d09a5ca73d41578cc604c9

ngimel commented:
Closing because internal tests were already fixed to accept Scalar, and I don't want to break xla again.
